keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,652 @@
+Instead, for such a use case, each controller, like [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md), implements its own logic to protect used objects from being deleted while in-use.
+This KEP aims to provide a generic mechanism for the use case, because such use cases are also found for protecting secrets and configmaps, [here](https://github.com/kubernetes/enhancements/pull/2640).
+
+Potential use cases are:
+- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
+- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- Protect resources that users marked as in-use
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validating admission webhook.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
+### Notes/Constraints/Caveats (Optional)
+
+- In some cases, users really wan't to delete the objects that are considered to be protected in general.
+For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.
```

Comment:
Since liens are not ACL'd, they just need to be manually removed to delete the object.